### ISSUE TITLE
Can`t intterupt vm.

### ISSUE THREAD

**Issue by iapon** [robertkrimen/otto#222]
```
package main

import (
	"errors"
	"fmt"
	"time"
	"github.com/robertkrimen/otto"
	"runtime"
)

var halt = errors.New("Stahp")

func main() {

	funcbody := `
		function create() {
    		var b = new Array(100000000).join('x');
			}
		create();
	`
	runtime.GOMAXPROCS(runtime.NumCPU())
	jsVM := otto.New()
	jsVM.Interrupt = make(chan func(), 1) 
	start := time.Now()
	defer func() {
		duration := time.Since(start)
		if caught := recover(); caught != nil {
			if caught == halt {
				fmt.Printf("Eval error: Code execution time exceeded limits. Stopping after: %v", duration)
				return
			}
			panic(caught) 
		}
	}()
	go func() {
		time.Sleep(3 * time.Second)
		jsVM.Interrupt <- func() {
			panic(halt)
		}
	}()
	fmt.Println("Starting eval")
	_, err := jsVM.Eval(funcbody)
	if err != nil {
		fmt.Printf("Eval error: %v", err)

	}
	fmt.Printf("Interrupt not work here bcz time for execution: %v", time.Since(start))
}

```

Prints 
```
C:\Projects\test_golang>go run main.go
Starting eval
Interrupt not work here bcz time for execution: 39.1490215s
```

How i can solve it?

**Comment by dhiltonp** [comment]
If you replace your funcbody with `while(true){}`, it works quite well.

The interrupt code isn't foolproof - jsVM.Interrupt is read from at just a couple of points in the execution engine.

**Comment by iapon** [comment]
Yes, I see it in  code. We need a mechanism that limits the memory usage. While I see only one way to avoid this particular implementation - check the length of the string in the resolve (), and since it is invoked at each iteration panic achievements of the threshold.  What do you think about this?

**Comment by dhiltonp** [comment]
The ability to limit memory usage would be a good feature.

Duktape supports this via custom memory allocators. That approach makes the most sense to me (adding a runtime.HaveMemory(size) to newString, newArray etc. in global.go), but we run into the problem of also needing to keep track of when memory is returned. If we're going down this rabbit hole, it might make sense to switch to fully manual memory management via something like https://github.com/couchbase/go-slab?

**Comment by iapon** [comment]
Or this: https://github.com/teh-cmc/mmm. 
Also we can try to use this article: https://blog.cloudflare.com/recycling-memory-buffers-in-go/
Anyway this is a complex problem and it does not have an easy solution.

**Comment by Niondir** [comment]
Actually I have issues with interrupting a endless for loop as well:

```
for(;;){
}
```

It's not able to interrupt this code. I see this rather critical.

**Comment by fratuz610** [comment]
I just put a PR in to fix this https://github.com/robertkrimen/otto/pull/374
